File: BIB/cocos2dx/gaf_runtimes/cocos2d-iphone/gaf/GAFActionObject.m
#import "GAFActionObject.h"
#import "GAFTextureAtlas.h"
#import "GAFTextureAtlasElement.h"
#import "ccMacros.h"


static NSString * const kClassNameKey   = @"className";
static NSString * const kNameKey        = @"name";
static NSString * const kXKey           = @"x";
static NSString * const kYKey           = @"y";
static NSString * const kRectXKey       = @"rectX";
static NSString * const kRectYKey       = @"rectY";
static NSString * const kRectWidthKey   = @"rectWidth";
static NSString * const kRectHeightKey  = @"rectHeight";


@interface GAFActionObject ()
@end

@implementation GAFActionObject

#pragma mark -
#pragma mark Properties

@synthesize name;
@synthesize pivotPoint;
@synthesize bounds;

#pragma mark -
#pragma mark Initialization & Release

- (id)initWithDictionary:(NSDictionary *)aDictionary
{
    self = [super init];
    if (nil != self)
    {
        self.name = [aDictionary objectForKey:kNameKey];
        
        NSNumber *x = [aDictionary objectForKey:kXKey];
        NSNumber *y = [aDictionary objectForKey:kYKey];
        if (x != nil && y != nil)
        {
            self.pivotPoint = CGPointMake([x integerValue], [y integerValue]);
        }
        else
        {
            CCLOGWARN(@"Error while creating GAFActionObject from NSDictionary. One of x, y attributes missing.");
            [self release];
            return nil;
        }
        
        NSNumber *rectX = [aDictionary objectForKey:kRectXKey];
        NSNumber *rectY = [aDictionary objectForKey:kRectYKey];
        NSNumber *rectWidth = [aDictionary objectForKey:kRectWidthKey];
        NSNumber *rectHeight = [aDictionary objectForKey:kRectHeightKey];
        if (rectX != nil &&
            rectY != nil &&
            rectWidth != nil &&
            rectHeight != nil)
        {
            self.bounds = CGRectMake([rectX integerValue], [rectY integerValue],
                                     [rectWidth integerValue], [rectHeight integerValue]);
        }
        else
        {
            CCLOGWARN(@"Error while creating GAFActionObject from NSDictionary. One of rect* attributes missing.");
            [self release];
            return nil;
        }
    }
    return self;
}

- (void)dealloc
{
    [name release];
    [super dealloc];
}

#pragma mark -
#pragma mark Public methods

#pragma mark -
#pragma mark Private methods

@end

File: BIBs/cocos2dx/MarmeladeGAFPlayer/Library/Sources/GAFActionObject.cpp
#include "GAFActionObject.h"

#include "cocoagaf/GAFInteger.h"
#include "cocoagaf/GAFDictionary.h"
#include "cocoagaf/GAFString.h"

static const char * kNameKey        = "name";
static const char * kXKey           = "x";
static const char * kYKey           = "y";
static const char * kRectXKey       = "rectX";
static const char * kRectYKey       = "rectY";
static const char * kRectWidthKey   = "rectWidth";
static const char * kRectHeightKey  = "rectHeight";

namespace GAF
{

GAFActionObject::GAFActionObject()
{
	
}

GAFActionObject * GAFActionObject::create(CCDictionary * aDictionary)
{
	GAFActionObject * ret = new GAFActionObject();
	if (ret && ret->initWithDictionary(aDictionary))
	{
		ret->autorelease();
		return ret;
	}
	CC_SAFE_DELETE(ret);
	return NULL;	
}

bool GAFActionObject::initWithDictionary(CCDictionary * aDictionary)
{
	if (!aDictionary)
	{
		GAFLOGERROR("Error while creating GAFActionObject from NSDictionary. aDictionary must not be NULL.");
		return false;
	}
	CCString * jName = (CCString*)aDictionary->objectForKey(kNameKey);
	if (jName)
	{
		name = std::string(jName->getCString());
	}
	
	CCInteger * rectX      = (CCInteger*)aDictionary->objectForKey(kRectXKey);
	CCInteger * rectY      = (CCInteger*)aDictionary->objectForKey(kRectYKey);
	CCInteger * rectWidth = (CCInteger*)aDictionary->objectForKey(kRectWidthKey);
	CCInteger * rectHeight  = (CCInteger*)aDictionary->objectForKey(kRectHeightKey);
	
	if (rectX && rectY && rectWidth && rectHeight)
	{
		bounds = GAFCCRectMake(rectX->getValue(), rectY->getValue(), rectWidth->getValue(), rectHeight->getValue());
	}
	else
	{
		GAFLOGERROR("Error while creating GAFActionObject from NSDictionary. One of rect* attributes missing.");
		return false;
	}
	
	CCInteger * x      = (CCInteger*)aDictionary->objectForKey(kXKey);
	CCInteger * y      = (CCInteger*)aDictionary->objectForKey(kYKey);
	
	if (x && y)
	{
		pivotPoint = GAFCCPointMake(x->getValue(), y->getValue());
	}
	else
	{
		GAFLOGERROR("Error while creating GAFActionObject from NSDictionary. One of x, y attributes missing.");
		return false;
	}
	return true;
}

} // namespace GAF
